### ISSUE TITLE
why use BCE loss instead of CCE loss for multi-class detection

### ISSUE THREAD

**Issue by captainst** [ultralytics/yolov5#5401]
## ❔Question
I see from the code in **ComputeLoss** that BCE (binary cross entropy) is used to calculate the cls and obj loss. For multi-class application, it seems that loss for multiple classes are broken down into multiple single class BCE. It confuses me. Shouldn't we use the CCE (categorical cross entropy) instead of BCE for multi-class application ?

Many thanks !

**Comment by glenn-jocher** [comment]
@captainst BCE loss provides better results and probably safer Classification balancing with Objectness which is also BCE. It also allows for YOLOv5 to be used in multi_label applications, i.e. 'flower' and 'rose'.

**Comment by glenn-jocher** [comment]
@captainst another point though is that this is an empirical science, so if you can produce better results with CE, let us know.

**Comment by captainst** [comment]
@glenn-jocher Thank you for your instant reply (that's fast!)
If I understood correctly, the BCE is designed also for multi-label apps, for example, "blue" and "shirt" for the same proposed box.
If, otherwise, in an application where each proposed box should be assigned to only on label (mutually exclusive), for example, bicycle, or people. In this case the CCE can also be used. Correct ?

**Comment by glenn-jocher** [comment]
@captainst yes that's right! CE can also be used with YOLOv5, though it would also require a rebalancing of the loss terms or an adjustment of the cls loss gain:
https://github.com/ultralytics/yolov5/blob/ed887b5976d94dc61fa3f7e8e07170623dc7d6ee/data/hyps/hyp.scratch.yaml#L14

**Comment by captainst** [comment]
@glenn-jocher many thanks again! I'll take a look at the yaml.

**Comment by eecn** [comment]
yolov5 uses BCEloss to achieve multi-classification including background classes (when all classes are less than the confidence value, such as 0.5, it is the background class), Why do you calculate obj_conf separately to judge whether to include samples or not? This confuses me.

**Comment by glenn-jocher** [comment]
@eecn dual obj and cls losses and outputs inherited from Joseph Redmon's original YOLOv3

**Comment by eecn** [comment]
> dual obj and cls losses and outputs inherited from Joseph Redmon's original YOLOv3
Thank you for your reply. I think this design is more convenient for the single label classification to replace BCE and CE, so as to maintain the consistency of post-processing.

**Comment by sujit-ag** [comment]
I've used CE loss instead of BCE with logits loss and saw huge performance improvement for my usecase. New set of experiments are in progress for training with higher gain value for ce loss and CE with focal loss.

**Comment by glenn-jocher** [comment]
@sujit-ag that's fantastic to hear! It's great to see you experimenting with different loss functions and observing performance improvements. Your feedback is valuable for the YOLO community. Keep up the great work!
